spark-cassandra-connector/src/main/scala/org/apache/spark/sql/cassandra/CassandraConfCache.scala

```diff
@@ -0,0 +1,204 @@
+  private val tableReadConf = java.util.Collections.synchronizedMap(
+    new java.util.HashMap[Seq[String], ReadConf]())
+
+  /** Add table level read configuration settings. Set cluster to None for a single cluster */
+  def addTableReadConf(keyspace: String,
+                       table: String,
+                       conf: ReadConf,
+                       cluster: Option[String]): Unit = {
+    cluster match {
+      case Some(c) => validateClusterName(c)
+        tableReadConf.put(Seq(table, keyspace, c), conf)
+      case _       => tableReadConf.put(Seq(table, keyspace), conf)
+    }
+  }
+
+  /** Add keyspace level read configuration settings. Set cluster to None for a single cluster */
+  def addKeyspaceLevelReadConf(keyspace: String,
+                               conf: ReadConf,
+                               cluster: Option[String]) : Unit = {
+    cluster match {
+      case Some(c) => validateClusterName(c)
+        keyspaceReadConf.put(Seq(keyspace, c), conf)
+      case _       => keyspaceReadConf.put(Seq(keyspace), conf)
+    }
+  }
+
+  /** Add cluster level read configuration settings */
+  def addClusterLevelReadConf(cluster: String, conf: ReadConf) : Unit = {
+    validateClusterName(cluster)
+    clusterReadConf.put(cluster, conf)
+  }
+
+  /** Remove table level read configuration settings */
+  def removeTableLevelReadConf(keyspace: String, table: String, cluster: Option[String]) : Unit = {
+    cluster match {
+      case Some(c) => validateClusterName(c)
+        tableReadConf.remove(Seq(table, keyspace, c))
+      case _       => tableReadConf.remove(Seq(table, keyspace))
+    }
+  }
+
+  /** Remove keyspace level read configuration settings */
+  def removeKeyspaceLevelReadConf(keyspace: String, cluster: Option[String]) : Unit = {
+    cluster match {
+      case Some(c) => validateClusterName(c)
+        keyspaceReadConf.remove(Seq(keyspace, c))
+      case _       => keyspaceReadConf.remove(Seq(keyspace))
+    }
+  }
+
+  /** Remove cluster level read configuration settings */
+  def removeClusterLevelReadConf(cluster: String) : Unit = {
+    clusterReadConf.remove(cluster)
+  }
+
+  /** Get read configuration settings by the order of table level, keyspace level, cluster level, default settings */
+  def getReadConf(keyspace: String,
+                  table: String,
+                  cluster: Option[String],
+                  defaultConf: ReadConf): ReadConf = {
+    cluster match {
+      case Some(c) => validateClusterName(c)
+        Option(tableReadConf.get(Seq(table, keyspace, c))).getOrElse(
+          Option(keyspaceReadConf.get(Seq(keyspace, c))).getOrElse(
+            Option(clusterReadConf.get(c)).getOrElse(defaultConf)))
+      case _       => Option(tableReadConf.get(Seq(table, keyspace))).getOrElse(
+        Option(keyspaceReadConf.get(Seq(keyspace))).getOrElse(defaultConf))
+    }
+  }
+
+  private val clusterWriteConf = java.util.Collections.synchronizedMap(
```

Comment:
DRY. Te same logic is repeated for `WriteConf` as well as for `ReadConf`. The type of the item stored in the cache can be abstracted away. Why not have a *single* generic cache with *two instances*, one for storing `ReadConf`s another for storing `WriteConf`s? And if we ever need another config "thing" to cache, you'd just add a single line of code to create another cache, instead of a third copy of the cache logic.